deepspeed/runtime/zero/stage_1_and_2.py

```diff
@@ -1308,7 +1308,8 @@ def complete_grad_norm_calculation_for_cpu_offload(self, params):
         total_norm = total_norm_cuda[0].item()**(1. / norm_type)
 
         if total_norm == float('inf') or total_norm == -float('inf') or total_norm != total_norm:
-            total_norm = -1
+            logger.info(f"Warning: invalid gradient detected. Please check your model implementation/configuration to improve the numerical stability.")
+            total_norm = -1.
```

Comment:
I think the idea here is to warn the user that this is an expected behavior and ususally it would cause/is caused by an error in the gradients. I'm fine with the current temporary fix, but I still hope we can probably throw an exception here or print some warning information to let the user know.

Reply:
Hi @tjruwase , yes I checked there is no subsequent overflow message warning. The code immediately crashed, after returning -1 here and then when calculating the grad norm elsewhere, PyTorch throws an uncaught exception when we are trying to calculate the norm of an (autocasted) torch.int64. I was using deepspeed directly with the official huggingface lm example code with mixed precision of bf16,  didn't change too much of it. For the error message, please refer to Issue #5241. It's the same in my case.

Reply:
@desire2020, thanks for the response. I think you have stumbled on a bigger problem in our code. Normally, we [check for overflows](https://github.com/microsoft/DeepSpeed/blob/b112c99ea8e09eb06ada0d60a3687983cb8c4bd0/deepspeed/runtime/zero/stage_1_and_2.py#L1820-L1821) before [computing gradient norms]. We assume that the conditions for an overflow are the same as for gradient norm of `-1`, and so norm computation is skipped on overflows.  For bf16 training, we don't check for [overflows](#4512) because we assume overflows are impossible. But it seems your test case contradicts our assumption. Can you try enabling [overflow checks for bf16 training](https://github.com/microsoft/DeepSpeed/blob/3dd3d514fbd964a1fbfe65880d6e4d13a5b8cd13/deepspeed/runtime/zero/stage_1_and_2.py#L1820-L1821) to see if overflow is detected? Thanks!

Reply:
Hi @tjruwase , I tested your suggestion. Yes enabling overflow checks for bf16 training does help to address the issue too. If you think this is a better fix, please include it in the next official release as we look forward to it.

P.S. I can explain why this problem could be occuring in my use case: I need a gradient to flow through a KL divergence term between two very peaky Bernoulli distributions, of which bf16 could be rounding some very small probability mistakenly to 0 more easily (since it has significantly reduced precision in exchange for larger range) and trigger the NaN error.

Reply:
@desire2020, thanks for this confirmation. Also, thanks for explanation how your scenario is causing bf16 overflows; that is great learning for me. 

I am actually conflicted on the right solution here. Avoiding overflow checks for bf16 is attractive because it works in most cases and avoids overflow checking overhead. However, it breaks in a few cases such as yours. 

How about this? Can you add a ds_config option [here](https://www.deepspeed.ai/docs/config-json/#bfloat16-training-options) to toggle overflow checking for bf16. This option would be disabled by default. This could look like: 

```json
  "bf16" : {
     ...
     "check_overflow": ["true" |  "false" ]
  }
```

Reply:
Hi @tjruwase , sorry if I'm asking a stupid question, is this a currently supported feature or we'd be expecting it in the next release of deepspeed? 'Cause I don't find related implementation of this feature in the current version.

Reply:
@desire2020, no worries, I was not clear. I was asking if you could add this feature to your PR? Thanks!

Reply:
More specifically, the desired feature is to add "check_overflow" option into the "bf16" dict of the ds_config.

Reply:
@tjruwase Ok, do you think I should add it for fp16 at the same time? If people are certain that their objective is stable and simple, they can use this option to further boost their fp16 mixed precision training too.

Reply:
Sure, but let's make the fp16 option true by default to preserve BC. Thanks!